### PR TITLE
DEVPROD-5762 Replace all hardcoded slugs with an enum slug and consolidate slug name to associated resource

### PR DESCRIPTION
**Pull request by khelif96**
DEVPROD-5762

### Description
This should not break any existing pages or urls and only serves to consolidate the slug names we use internally in the Spruce application. 
- [x] Create Slugs for each existing hardcoded slug.
- [x] Update generic slugs so they correspond with the associated resource.
- [x] Update the usage of hardcoded slugs with their enum slug equivalent

### Testing
Existing tests should suffice to catch any edge cases that could have been missed.

### CODE REVIEW

File: src/pages/Version.tsx [hunk 1]
```diff
@@ -38,7 +38,7 @@ import { NameChangeModal } from "./version/NameChangeModal";
 // docs/decisions/2023-12-13_version_page_logic.md
 export const VersionPage: React.FC = () => {
   const spruceConfig = useSpruceConfig();
-  const { id } = useParams<{ id: string }>();
+  const { [slugs.versionId]: id } = useParams();
```
Comment (minnakt) [inline src/pages/Version.tsx:41]:
```suggestion
  const { [slugs.versionId]: versionId } = useParams();
```

File: src/pages/host/index.tsx [hunk 2]
```diff
@@ -33,7 +34,7 @@ const { getLimitFromSearch, getPageFromSearch } = url;
 
 const Host: React.FC = () => {
   const dispatchToast = useToastContext();
-  const { id } = useParams<{ id: string }>();
+  const { [slugs.hostId]: id } = useParams();
```
Comment (minnakt) [inline src/pages/host/index.tsx:37]:
```suggestion
  const { [slugs.hostId]: hostId } = useParams();
```

File: src/pages/projectSettings/index.tsx [hunk 3]
```diff
@@ -36,10 +37,11 @@ const { validateObjectId } = validators;
 const ProjectSettings: React.FC = () => {
   usePageTitle(`Project Settings`);
   const dispatchToast = useToastContext();
-  const { projectIdentifier: identifier, tab } = useParams<{
-    projectIdentifier: string;
-    tab: ProjectSettingsTabRoutes;
-  }>();
+  const { [slugs.projectIdentifier]: identifier, [slugs.tab]: tab } =
+    useParams<{
+      [slugs.projectIdentifier]: string | null;
```
Comment (minnakt) [inline src/pages/projectSettings/index.tsx:42]: just curious if `| null` is required here:
```suggestion
      [slugs.projectIdentifier]: string;
```

File: src/pages/taskQueue/index.tsx [hunk 4]
```diff
@@ -27,7 +27,7 @@ import TaskQueueTable from "./TaskQueueTable";
 const TaskQueue = () => {
   const taskQueueAnalytics = useTaskQueueAnalytics();
 
-  const { distro } = useParams<{ distro: string }>();
+  const { [slugs.distro]: distro } = useParams();
```
Comment (minnakt) [inline src/pages/taskQueue/index.tsx:30]: do you think we should just make TaskQueue use the `distroId` slug?

File: src/pages/version/VersionTabs.tsx [hunk 5]
```diff
@@ -35,14 +35,14 @@ const tabMap = ({
   patchId: string;
 }) => ({
   [PatchTab.Tasks]: (
-    <Tab name="Tasks" id="task-tab" data-cy="task-tab" key="tasks-tab">
+    <Tab name="Tasks" versionId="task-tab" data-cy="task-tab" key="tasks-tab">
```
Comment (minnakt) [inline src/pages/version/VersionTabs.tsx:38]: the updates to `id` on the `<Tab />` component seem like a mistake (4 instances in this file)

File: src/pages/version/VersionTabs.tsx [hunk 6]
```diff
@@ -86,9 +86,9 @@ export const VersionTabs: React.FC<Props> = ({
   isPatch,
   taskCount,
 }) => {
-  const { id, tab } = useParams<{ id: string; tab: PatchTab }>();
+  const { [slugs.versionId]: versionId, [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/pages/version/VersionTabs.tsx:89]: we lose some type specificity without `tab: PatchTab`; maybe want to put it back?

File: src/hooks/useConfigurePatch.ts [hunk 7]
```diff
@@ -152,7 +152,7 @@ export const useConfigurePatch = (
 ): HookResult => {
   const navigate = useNavigate();
   const location = useLocation();
-  const { tab } = useParams<{ tab: PatchTab | null }>();
+  const { [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/hooks/useConfigurePatch.ts:155]: same comment

File: src/pages/configurePatch/configurePatchCore/useConfigurePatch/index.ts [hunk 8]
```diff
@@ -125,7 +125,7 @@ interface HookResult extends ConfigurePatchState {
 const useConfigurePatch = (patch: ConfigurePatchQuery["patch"]): HookResult => {
   const navigate = useNavigate();
   const location = useLocation();
-  const { tab } = useParams<{ tab: PatchTab | null }>();
+  const { [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/pages/configurePatch/configurePatchCore/useConfigurePatch/index.ts:128]: same comment

File: src/pages/preferences/PreferencesTabs.tsx [hunk 9]
```diff
@@ -10,7 +14,7 @@ import { ProfileTab } from "./preferencesTabs/ProfileTab";
 import { PublicKeysTab } from "./preferencesTabs/PublicKeysTab";
 
 export const PreferencesTabs: React.FC = () => {
-  const { tab } = useParams<{ tab: string }>();
+  const { [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/pages/preferences/PreferencesTabs.tsx:17]: can add more type specificity with `<{ tab: PreferencesTabRoutes }>`

File: src/pages/preferences/index.tsx [hunk 10]
```diff
@@ -7,13 +7,17 @@ import {
...
+  getPreferencesRoute,
+  slugs,
+} from "constants/routes";
 import { usePageTitle } from "hooks";
 import { PreferencesTabs } from "pages/preferences/PreferencesTabs";
 
 const Preferences: React.FC = () => {
   usePageTitle("Preferences");
-  const { tab } = useParams<{ tab: string }>();
+  const { [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/pages/preferences/index.tsx:20]: same comment

File: src/pages/spawn/index.tsx [hunk 11]
```diff
@@ -6,12 +6,12 @@ import {
...
   SideNavItem,
 } from "components/styles";
-import { routes, SpawnTab } from "constants/routes";
+import { routes, SpawnTab, slugs } from "constants/routes";
 import { SpawnHost } from "./SpawnHost";
 import { SpawnVolume } from "./SpawnVolume";
 
 const Spawn: React.FC = () => {
-  const { tab } = useParams<{ tab: string }>();
+  const { [slugs.tab]: tab } = useParams();
```
Comment (minnakt) [inline src/pages/spawn/index.tsx:14]: can add more type specificity with `<{ tab: SpawnTab }>`

File: src/pages/version/VersionTabs.tsx [hunk 12]
```diff
@@ -109,9 +109,9 @@ export const VersionTabs: React.FC<Props> = ({
       taskCount,
       childPatches,
       numFailedChildPatches,
-      patchId: id,
+      patchId: versionId,
```
Comment (minnakt) [inline src/pages/version/VersionTabs.tsx:112]: changing this parameter to be called `versionId` might make this a bit less confusing to read

File: src/constants/routes.ts [hunk 13]
```diff
@@ -74,38 +74,56 @@ const paths = {
...
-  distroSettings: `${paths.distro}/:distroId/${PageNames.Settings}`,
-  host: `${paths.host}/:id`,
+  configurePatch: `${paths.patch}/:${slugs.patchId}/configure`,
+  container: `${paths.container}/:${slugs.podId}`,
+  distroSettings: `${paths.distro}/:${slugs.distroId}/${PageNames.Settings}`,
+  host: `${paths.host}/:${slugs.hostId}`,
   hosts: paths.hosts,
-  jobLogs: `${paths.jobLogs}/:buildId`,
+  jobLogs: `${paths.jobLogs}/:${slugs.buildId}`,
+  jobLogsGroup: `${paths.jobLogs}/:${slugs.taskId}`,
```
Comment (minnakt) [inline src/constants/routes.ts:110]: where does this one come from? is `taskId` the correct parameter rather than something like `groupId`?

File: src/constants/routes.ts [hunk 14]
```diff
@@ -74,38 +74,56 @@ const paths = {
...
   login: paths.login,
   myPatches: `${paths.user}/${PageNames.Patches}`,
-  patch: `${paths.patch}/:id`,
+  patch: `${paths.patch}/:${slugs.versionId}`,
```
Comment (minnakt) [inline src/constants/routes.ts:113]: should it be `patchId`? (not sure, just guessing since it seems odd)

**Review by minnakt — CHANGES_REQUESTED** [review]

File: src/pages/taskQueue/index.tsx (hunk 4 above)
Comment (khelif96) [reply src/pages/taskQueue/index.tsx:30]: Good idea we can consolidate them

File: src/pages/version/VersionTabs.tsx (hunk 5 above)
Comment (khelif96) [reply src/pages/version/VersionTabs.tsx:38]: Good eye 💯

File: src/constants/routes.ts (hunk 13 above)
Comment (khelif96) [reply src/constants/routes.ts:121]: Good catch this is from my job logs pr I must have mixed up the commits

File: src/constants/routes.ts (hunk 14 above)
Comment (khelif96) [reply src/constants/routes.ts:124]: Woah this page should not exist it looks like we have 2 version pages with 2 urls

File: src/hooks/useLegacyUIURL/index.ts [hunk 15]
```diff
@@ -45,3 +40,19 @@ export const useLegacyUIURL = (): string | null => {
 
   return legacyUIUrl;
 };
+
+/**
+ * `slugToId` is a helper function that takes a Params object from react-router-dom's matchPath and returns a slug value from the slugs object if it is a suitable id.
+ * @param matchedPathParams - Params object from react-router-dom's matchPath
+ * @returns string | undefined
+ */
+const slugToId = (matchedPathParams: Params): string | undefined => {
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/hooks/useLegacyUIURL/index.ts:49]: Good catch on this I updated it so that non id slugs are no longer included in the check. I preferred this approach over the previous implementation of just casing on every single slug.

File: src/pages/configurePatch/configurePatchCore/useConfigurePatch/index.ts (hunk 8 above)
Comment (minnakt) [reply src/pages/configurePatch/configurePatchCore/useConfigurePatch/index.ts:128]: bump!

File: src/constants/routes.ts [hunk 16]
```diff
@@ -74,38 +74,66 @@ const paths = {
...
+  projectSettings: `${paths.project}/:${slugs.projectIdentifier}/${PageNames.Settings}`,
   spawn: paths.spawn,
   spawnHost: `${paths.spawn}/${SpawnTab.Host}`,
   spawnVolume: `${paths.spawn}/${SpawnTab.Volume}`,
-  task: `${paths.task}/:id`,
-  taskHistory: `${paths.taskHistory}/:projectIdentifier/:taskName`,
-  taskQueue: paths.taskQueue,
+  task: `${paths.task}/:${slugs.taskId}`,
+  taskHistory: `${paths.taskHistory}/:${slugs.projectIdentifier}/:${slugs.taskName}`,
+  taskQueue: `${paths.taskQueue}`,
```
Comment (minnakt) [inline src/constants/routes.ts:132]:
```suggestion
  taskQueue: paths.taskQueue},
```

File: src/constants/routes.ts (hunk 14 above)
Comment (minnakt) [reply src/constants/routes.ts:123]: (i saw you tried deleting it but it didn't work, just curious what you found out this page is for)

**Review by minnakt — APPROVED** [review]

**Comment by minnakt** [comment]
Closing this PR since it has already been moved to the UI repo
